go-selinux/selinux_linux.go

```diff
@@ -445,6 +469,194 @@ func ComputeCreateContext(source string, target string, class string) (string, e
+
+	switch len(levelSlice) {
+	case 2:
+		high := strings.SplitN(levelSlice[1], ":", 2)
+		sensHigh, err := parseLevelItem(high[0], sensitivity)
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.high = &level{sens: sensHigh}
+		if len(high) > 1 {
+			catsHigh, err := catsToBitset(high[1])
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse categories")
+			}
+			mlsRange.high.cats = catsHigh
+		}
+		fallthrough
+	case 1:
+		low := strings.SplitN(levelSlice[0], ":", 2)
+		sensLow, err := parseLevelItem(low[0], sensitivity)
+		if err != nil {
+			return nil, err
+		}
+		mlsRange.low = &level{sens: sensLow}
+		if len(low) > 1 {
+			catsLow, err := catsToBitset(low[1])
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to parse categories")
+			}
+			mlsRange.low.cats = catsLow
+		}
+	}
+
+	if mlsRange.high == nil {
+		mlsRange.high = mlsRange.low
+	}
+
+	return mlsRange, nil
+}
+
+/*
+bitsetToStr takes a category bitset and returns it in the
+canonical selinux syntax
+*/
+func bitsetToStr(c *bitset.BitSet) string {
+	var catBuff bytes.Buffer
+	i, e := c.NextSet(0)
+	len := 0
+	for e {
+		if len == 0 {
+			if catBuff.Len() > 0 {
+				catBuff.WriteString(",")
+			}
+			catBuff.WriteString("c" + strconv.FormatInt(int64(i), 10))
+		}
+
+		next, e := c.NextSet(i + 1)
+		if e {
+			// consecutive cats
+			if next == i+1 {
+				len++
+				i = next
+				continue
+			}
+		}
+		if len == 1 {
+			catBuff.WriteString(",c" + strconv.FormatInt(int64(i), 10))
+		} else if len > 1 {
+			catBuff.WriteString(".c" + strconv.FormatInt(int64(i), 10))
+		}
+		if !e {
+			break
+		}
+		len = 0
+		i = next
+	}
+
+	return catBuff.String()
+}
+
+// String returns an mlsRange as a string.
+func (m mlsRange) String() string {
+	low := fmt.Sprintf("s%d", m.low.sens)
+	if m.low.cats != nil && m.low.cats.Count() > 0 {
+		low += ":" + bitsetToStr(m.low.cats)
+	}
+
+	if reflect.DeepEqual(m.low, m.high) {
+		return low
+	}
+
+	high := fmt.Sprintf("s%d", m.high.sens)
+	if m.high.cats != nil && m.high.cats.Count() > 0 {
+		high += ":" + bitsetToStr(m.high.cats)
+	}
+
+	return low + "-" + high
+}
+
+/*
+CalculateGlbLub computes the glb (greatest lower bound) and lub (least upper bound)
+of a source and target range.
+The glblub is calculated as the greater of the low sensitivities and
+the lower of the high sensitivities and the and of each category bitset.
+*/
+func CalculateGlbLub(sourceRange, targetRange string) (string, error) {
+	s, err := rangeStrToMLSRange(sourceRange)
+	if err != nil {
+		return "", err
+	}
+	t, err := rangeStrToMLSRange(targetRange)
+	if err != nil {
+		return "", err
+	}
+
+	if s.high.sens < t.low.sens || t.high.sens < s.low.sens {
+		/* these ranges have no common sensitivities */
+		return "", ErrIncomparable
+	}
+
+	outrange := &mlsRange{low: &level{}, high: &level{}}
+
+	/* take the greatest of the low */
+	outrange.low.sens = uint32(math.Max(float64(s.low.sens), float64(t.low.sens)))
```

Comment:
The `int` to `float` (and back) conversion just to use `Max()` looks unholy to me. It's probably very slow and prone to errors.

Until we have generics or `?:` operator in Go, I think in this case it would be the best to write your own min and max functions.

Reply:
👍 will do